### PR TITLE
adds armor absorption, where armor absorbs a portion of the blow to you.

### PR DESCRIPTION
**Pull request by MadmanMartian**
the current armor system is two dice rolls of armor negation. First one succeeding means the damage received is half, second one succeeding means the damage is fully negated.

armor_absorb is how much of the final damage is negated (by taking the damage to be received and dividing it against armor_absorb, then multipling the damage by the variable given should it be below 1), and is done after the two dicerolls so as to be more effective.

### **Example situation**

You're hit with a bullet that would normally deal 30 damage, fired from the clown's bananagun. Lucky for you, you're wearing GTide's patented armored grey jumpsuit, with 25 bullet armor negation, and 25 bullet armor absorb.

When the bullet strikes, you roll out of 100. If you get 25 or more, the damage dealt is 15. You roll again, and if it's 25 or more, the damage dealt is 0. However in your situation, you're only so much lucky, only negating the bullets damage to 15 thanks to the power of GTide's armor dispersal system (Clumps of newspapers sealed together using a combination of welder fuel and strange moss found in station maintenance)

Now the thickness of the jumpsuit is taken into account of how much energy the bullet loses on its way to your abdomen beneath. The 15 damage bullet is now divided against the bullet armor absorb, 15/25 = 0.6. 15*0.6 = 9. 9 damage is dealt to your abdomen as the bullet tickles you.

Now, if it were the inverse (armor absorb calculated before armor negation) it would go something like this.

30/25 = 1.2, so the armor_absorb does jack squat in this case. The armor negation rolls succesful once as it does in the above example, reducing the damage down to 15.

9 damage received if AN before AA, 15 damage received if AA before AN.

### **TL;DR**

AN is probability out of 100 you halve the damage, and a further probability out of 100 you negate the damage entirely.

AA is how much of the damage is absorbed. If damage is greater than AA then it does nothing, otherwise it reduces the incoming damage by damage*(damage/AA)

Example: 30 incoming damage, 1 succesful roll of AN, 25 AA

_**If AA before AN**_

30/25 > 1, no damage absorbed, still 30.

1 succesful roll of AA halves damage, damage now 15.

15 damage received.

_**If AN before AA**_

1 succesful roll of AN halves damage, damage now 15.

15/25 < 1, damage absorbed. 15*(15/25) = 9.

9 damage received.

### **TL;DR of TL;DR**

AN before AA > AA before AN

**I can change it to be the opposite case should the majority feel it should be that way.**

### CODE REVIEW

**Comment by yclatious** [comment]
Can you provide any numbers math for things like:
-Ablative vs normal las/las cannon
-Bullet Proof Vest vs Shotgun slug/Revolver
-Knight Armor vs Esword
Just want to check balance on it before I say yes, seems like a good change but if a ablative allows you to essentialy tank a las/pulse rifle or some knight armor almost nulls toolboxes or something else thats silly then maybe it needs some touching up in the numbers department.

**Comment by MadmanMartian** [comment]
Currently armor absorption's not implemented for anything. No PoC, just the system with which to use.

**Comment by yclatious** [comment]
This is what I get for sleeping 6 hours, making a moron out of myself.

If its just the system and no balance is actually being discussed then why the fuck not.

**Comment by Blithering** [comment]
This is neat, and could make the less powerful armours more worthwhile

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 1]
```diff
@@ -76,6 +94,21 @@ emp_act
...
 
+/mob/living/carbon/human/proc/checkarmorabsorb(var/datum/organ/external/def_zone, var/type)
+	if(!type)
+		return 0
+	var/protection = 0
+	var/list/body_parts = list(head, wear_mask, wear_suit, w_uniform, gloves, shoes)
+	for(var/bp in body_parts)
+		if(!bp)
+			continue
+		if(bp && istype(bp ,/obj/item/clothing))
```
Comment (DamianX) [inline code/modules/mob/living/carbon/human/human_defense.dm]: at least the first check on the this line is redundant, you could also get rid of the one above because `istype` will fail for null values. Also dat comma.

File: code/modules/mob/living/living_defense.dm [hunk 2]
```diff
@@ -39,6 +39,20 @@
 /mob/living/proc/getarmor(var/def_zone, var/type)
 	return 0
 
+/mob/living/proc/getarmorabsorb(var/def_zone, var/type)
```
Comment (DamianX) [inline code/modules/mob/living/living_defense.dm:42]: `get_armor_absorb` maybe?

File: code/modules/mob/living/carbon/human/human_defense.dm (hunk 1 above)
Comment (PJB3005) [reply code/modules/mob/living/carbon/human/human_defense.dm]: Yeah, both the `!bp` and `bp &&` here are redundant. Remove them please. And fix that comma.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

**Comment by Killette2** [comment]
This sounds cool on paper but I'm curious how much it would change combat

**Comment by MadmanMartian** [comment]
damage reduction

**Comment by Blithering** [comment]
It would make the 'softened blow' effect more useful, rather than just reducing damage by a flat 50%. It could also make armours more varied, an armour with a low absorb amount but high chance to block would be a riskier option than one with a high absorb, but potentially higher payoff
